### PR TITLE
Fixed positional embedding in blanx

### PR DESCRIPTION
**Pull request by ciebson123**
# Description
Added new way of telling "which token is on what position", which includes blanks. Now, all tokens after blanks see their positions as if there were no blanks before them. The exception is blanks themself, which see all their position including all blanks in the same sequence.

Example (b is blank).
If the sequence is   aabaabba
Then positions are 01223454.

# Checklist
   - [x] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [x] I attached link to neptune if it was necessary
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [x] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [x] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [x] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [x] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

File: research/blanks/configs/scratchpad/eval.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+parent: research/conditional/train/configs/baselines/gpt/small.yaml
+md5_parent_hash: cbdf4e5992f41713488c25c819086e51
+time: "0-20:10:00"
```
Comment (sebastianjaszczur) [inline research/blanks/configs/scratchpad/eval.yaml:3]: 20h, but with an additional 10m? Seems like hitting a "snooze" button in the morning.

Comment (maciek-pioro) [reply research/blanks/configs/scratchpad/eval.yaml:3]: xD

File: research/blanks/model.py [hunk 2]
```diff
@@ -221,3 +273,48 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+            fan_in=1,
+            init_type=init_type,
+            scale=init_scale,
+            dtype=default_weight.dtype,
+        )
+        self.blank_token_id = blank_token_id
+
+    def forward(self, x):
+        # stack overflow https://stackoverflow.com/questions/18196811/cumsum-reset-at-nan
+        positions = torch.arange(0, x.shape[-1], device=x.device)
```
Comment (maciek-pioro) [inline research/blanks/model.py:301]: Let's unsqueeze positions before multiplying, so broadcasting is easier to think about

File: research/blanks/model.py [hunk 3]
```diff
@@ -221,3 +273,48 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        positions = positions * torch.ones_like(x)
+        blank_mask = x.eq(self.blank_token_id)
```
Comment (maciek-pioro) [inline research/blanks/model.py:303]: I prefer `is_blank` and `is_not_blank`, but that's up to you

File: research/blanks/model.py [hunk 4]
```diff
@@ -221,3 +273,48 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        n_blanks_before = blank_mask.cumsum(dim=-1)
```
Comment (maciek-pioro) [inline research/blanks/model.py:304]: `n_blanks_up_to`? Before suggests not including the current position, but we include the current position.

File: research/blanks/model.py [hunk 5]
```diff
@@ -221,3 +273,48 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        positions = (
+            positions - n_blanks_before
+        )  # adjecent blanks have the same position
```
Comment (maciek-pioro) [inline research/blanks/model.py:307]: adjecent -> adjacent

File: research/blanks/model.py [hunk 6]
```diff
@@ -221,3 +273,48 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+        not_blank_mask = ~blank_mask
+        num_adjecent_blanks = torch.diff(
```
Comment (maciek-pioro) [inline research/blanks/model.py:309]: adjecent -> adjacent

File: research/blanks/model.py [hunk 7]
```diff
@@ -221,3 +273,48 @@ def forward(self, x: torch.Tensor) -> torch.Tensor:
...
+            n_blanks_before[not_blank_mask],
+            dim=-1,
+            prepend=torch.zeros_like(not_blank_mask[:, :1]),
+        )  # for each non-blank token, how many blanks are immediately before it
+        blank_mask[
+            not_blank_mask
+        ] = -num_adjecent_blanks  # now it looks like [0, 1, 1, -2, 0, 1, 1, 1, -3, ..]
+        fixup = blank_mask.cumsum(dim=-1)  # [0, 1, 2, 0, 0, 1, 2, 3, 0, ..]
+        positions = positions + fixup
```
Comment (maciek-pioro) [inline research/blanks/model.py:318]: Let's extract all of that and let's add a test confirming that it works like we'd like it to

**Review by maciek-pioro — CHANGES_REQUESTED** [review]
Added a couple of comments. Good work !!

File: research/blanks/model.py (hunk 6 above)
Comment (ciebson123) [reply research/blanks/model.py]: Nit, but done

File: research/blanks/model.py (hunk 4 above)
Comment (ciebson123) [reply research/blanks/model.py]: Done

File: research/blanks/model.py (hunk 3 above)
Comment (ciebson123) [reply research/blanks/model.py]: Done for consistency in repo

File: research/blanks/model.py (hunk 2 above)
Comment (ciebson123) [reply research/blanks/model.py]: Outdated, now we have loop implementation :(

File: research/blanks/configs/scratchpad/eval.yaml (hunk 1 above)
Comment (ciebson123) [reply research/blanks/configs/scratchpad/eval.yaml:3]: Will disappear after merge of blank evaluation xd

**Comment by ciebson123** [comment]
@maciek-pioro added comments and tests

**Review by maciek-pioro — CHANGES_REQUESTED** [review]
I feel we can vectorize that.

File: research/blanks/utils.py [hunk 8]
```diff
@@ -66,3 +86,64 @@ def iterate_through_nth_blanks_masks(
...
+    return blank_insertion_point <= get_last_point_to_fit_blanks(
+        sequence_length, n_blanks
+    )
+
+
+def make_blanks_fixed_positions(x: torch.Tensor, blank_token_id: int):
+    # stack overflow https://stackoverflow.com/questions/18196811/cumsum-reset-at-nan
+    # let's go with example x = [0, 1, 1, 0, 1] - 1 is blank
+    final_positions = torch.zeros_like(x)
+    for i, example in enumerate(x):
```
[the remark this answers is not in the document]
Comment (ciebson123) [reply research/blanks/utils.py:127]: We could make it much simpler and vectorized if we assume that each blank sequence has exactly n_blanks

**Comment by maciek-pioro** [comment]
Let's do that! And if that's easy to check, let's put that into an assert.

**Comment by ciebson123** [comment]
I will add the assert that total number of blanks is divisible by number of blanks per block, should be close enough

File: research/blanks/utils.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ciebson123) [reply research/blanks/utils.py]: Done

**Comment by ciebson123** [comment]
https://app.neptune.ai/o/pmtest/org/llm-random/runs/table?viewId=9aa26157-4023-4c29-a82f-3e83983a82b9&compare=MwFhAYBoEYg

**Review by maciek-pioro — APPROVED** [review]

**Comment by maciek-pioro** [comment]
LGTM!!!

**Pull request merged**